### PR TITLE
add 32 tube carrier for 1.5ml eppendorf tubes

### PR DESCRIPTION
**Pull request by ben-ray**
While exploring with the TubeCarrier objects, I noticed something strange:

Here is how I define & use TubeCarriers:

```
tube_car = lh.get_resource("tube_car")
lh.deck.assign_child_resource(resource=Tube_CAR_32_A00(name="tube_car"), rails=5, reassign=True, replace=True)
for i in range(8):
    tube_car [i] = Falcon_tube_14mL_Rb(f"tube_{i}")
await lh.pick_up_tips()
await lh.aspirate([tube_car[i].resource for i in range(8)][::-1], [0]*8) 
```
notice the extra .resource in the aspirate command? if I don't use this, I get the following error:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[47], [line 5](vscode-notebook-cell:?execution_count=47&line=5)
      [2](vscode-notebook-cell:?execution_count=47&line=2) from pylabrobot.resources.tip_tracker import set_tip_tracking
      [3](vscode-notebook-cell:?execution_count=47&line=3) from pylabrobot.resources.volume_tracker import set_volume_tracking
----> [5](vscode-notebook-cell:?execution_count=47&line=5) await lh.aspirate([tube_car[i] for i in range(8)][::-1], [0]*8) 

File c:\users\ubeast\desktop\ben\wetlab_workcell\src\pylabrobot\pylabrobot\machines\machine.py:25, in need_setup_finished.<locals>.wrapper(self, *args, **kwargs)
     [23](file:///C:/users/ubeast/desktop/ben/wetlab_workcell/src/pylabrobot/pylabrobot/machines/machine.py:23) if not self.setup_finished:
     [24](file:///C:/users/ubeast/desktop/ben/wetlab_workcell/src/pylabrobot/pylabrobot/machines/machine.py:24)   raise RuntimeError("The setup has not finished. See `setup`.")
---> [25](file:///C:/users/ubeast/desktop/ben/wetlab_workcell/src/pylabrobot/pylabrobot/machines/machine.py:25) return await func(self, *args, **kwargs)

File c:\users\ubeast\desktop\ben\wetlab_workcell\src\pylabrobot\pylabrobot\liquid_handling\liquid_handler.py:723, in LiquidHandler.aspirate(self, resources, vols, use_channels, flow_rates, offsets, liquid_height, blow_out_air_volume, **backend_kwargs)
    [718](file:///C:/users/ubeast/desktop/ben/wetlab_workcell/src/pylabrobot/pylabrobot/liquid_handling/liquid_handler.py:718) assert len(resources) == len(vols) == len(offsets) == len(flow_rates) == len(liquid_height)
    [720](file:///C:/users/ubeast/desktop/ben/wetlab_workcell/src/pylabrobot/pylabrobot/liquid_handling/liquid_handler.py:720) # If the user specified a single resource, but multiple channels to use, we will assume they
    [721](file:///C:/users/ubeast/desktop/ben/wetlab_workcell/src/pylabrobot/pylabrobot/liquid_handling/liquid_handler.py:721) # want to space the channels evenly across the resource. Note that offsets are relative to the
    [722](file:///C:/users/ubeast/desktop/ben/wetlab_workcell/src/pylabrobot/pylabrobot/liquid_handling/liquid_handler.py:722) # center of the resource.
--> [723](file:///C:/users/ubeast/desktop/ben/wetlab_workcell/src/pylabrobot/pylabrobot/liquid_handling/liquid_handler.py:723) if len(set(resources)) == 1:
    [724](file:///C:/users/ubeast/desktop/ben/wetlab_workcell/src/pylabrobot/pylabrobot/liquid_handling/liquid_handler.py:724)   resource = resources[0]
    [725](file:///C:/users/ubeast/desktop/ben/wetlab_workcell/src/pylabrobot/pylabrobot/liquid_handling/liquid_handler.py:725)   n = len(use_channels)

TypeError: unhashable type: 'CarrierSite'
```

something I should tidy up before merging. any ideas what might be going wrong?

### CODE REVIEW

**Comment by rickwierenga** [comment]
it seems in the first two lines

```python
tube_car = lh.get_resource("tube_car")
lh.deck.assign_child_resource(resource=Tube_CAR_32_A00(name="tube_car"), rails=5, reassign=True, replace=True)
```

that the `Tube_CAR_32_A00` object is only made after you assign the variable `tube_car` (`tube_car` likely refers to a CarrierSite, not the tube carrier itself.) Normally, you would get an error because apparently there are two resources with the same name (this is a given, otherwise `.get_resource` would raise ResourceNotFoundError), but the error is not raised because `reassign=True`.

Now because `tube_car` is a `CarrierSite`, its child resource is accessed via `.resource` (or `.children[0]`). I think switching those first two lines would fix that.

File: pylabrobot/resources/carrier.py [hunk 1]
```diff
@@ -27,14 +29,22 @@ def assign_child_resource(
...
     self.resource = resource
-    return super().assign_child_resource(resource, location=location)
+    return super().assign_child_resource(resource, location, reassign)
 
   def unassign_child_resource(self, resource):
     self.resource = None
     return super().unassign_child_resource(resource)
+  
+  def serialize(self) -> dict:
+      return { **super().serialize(), "name": self.name}
```
Comment (rickwierenga) [inline pylabrobot/resources/carrier.py:39]: "name" shoudl already be serialized in super, so this function does not need overriding.

File: pylabrobot/resources/carrier.py [hunk 2]
```diff
@@ -27,14 +29,22 @@ def assign_child_resource(
...
+  
+  def __hash__(self):
+        return hash((self.name))
```
Comment (rickwierenga) [inline pylabrobot/resources/carrier.py:43]: is this needed over `Resource.__hash__`?

```python
  def __hash__(self) -> int:
    return hash(repr(self))
```

File: pylabrobot/resources/carrier.py [hunk 3]
```diff
@@ -276,7 +287,31 @@ def __init__(
...
-      sites,category=category, model=model)
+      sites,category=category, model=model)\
+    
+    def assign_child_resource(self, resource, location, reassign=False):
+      if not isinstance(resource, CarrierSite):
+          print(f"Type of resource: {type(resource)}")
+          print(f"Resource details: {resource}")
+          raise TypeError(f"Invalid resource {resource}")
+      self.sites.append(resource)
+      super().assign_child_resource(resource, location=location)
```
Comment (rickwierenga) [inline pylabrobot/resources/carrier.py:298]: `Carrier.assign_child_resource` has this already. I don't immediately see what's different here, but if I missed something, could you fix it in `Carrier` so it's fixed for all of them?

**Review by rickwierenga — COMMENTED** [review]
mostly methods were overriden by children where this was not necessary, presumably debugging artifacts or because of the issue you described in your first message in this thread.

PR is very much worth keeping open for the Eppendorf tube and `Tube_CAR_32_A00` additions

File: pylabrobot/resources/carrier.py (hunk 2 above)
Comment (ben-ray) [reply pylabrobot/resources/carrier.py:43]: will revert this, holdover from debugging.
apologies for the messy PR - realized I was stuck & you found exactly why, thank you!

**Comment by rickwierenga** [comment]
please mark as ready if you think it's good to merge this

**Comment by ben-ray** [comment]
had to get the coordinates more exactly correct. this current def has the z-bottom calibrated to where a 1.5ml eppendorf tube bottom will land, but more work needs to be done to Tube to ensure different z offsets get applied for different tube types (hanging vs non-hanging...)

**Pull request merged**